### PR TITLE
test(x/ecocredit): basket create acceptance tests

### PR DESCRIPTION
**Pull request by ryanchristo**
## Description

Closes: #895

This pull request updates and implements basket create acceptance tests.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/ecocredit/basket/date_criteria.go [hunk 1]
```diff
@@ -17,3 +19,34 @@ func (d *DateCriteria) ToApi() *api.DateCriteria {
 	}
 	return nil
 }
+
+// Validate validates DateCriteria, ensuring only one field is set,
+// minimum start date is after 1900-01-01, start date window is at
+// least one day, and years in the past is at least one year
+func (d *DateCriteria) Validate() error {
```
Comment (ryanchristo) [inline x/ecocredit/basket/date_criteria.go:26]: Moved from `msg_create.go` to `date_criteria.go` without changes to the underlying logic.

File: x/ecocredit/basket/date_criteria_test.go [hunk 2]
```diff
@@ -29,3 +30,82 @@ func TestDateCriteriaToApi(t *testing.T) {
 	require.NotNil(dw)
 	require.Equal(durStd, dw.AsDuration(), "handles window date")
 }
+
+func TestValidateDateCriteria(t *testing.T) {
```
Comment (ryanchristo) [inline x/ecocredit/basket/date_criteria_test.go:34]: Moved from `msg_create_test.go` to `date_criteria_test.go` without changes the test cases.

File: x/ecocredit/basket/utils.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import (
+	"github.com/regen-network/regen-ledger/x/ecocredit/core"
+)
+
+const basketDenomPrefix = "eco"
+
+// FormatBasketDenom formats denom and display denom:
+// - denom: eco.<m.Exponent><m.CreditTypeAbbrev>.<m.Name>
+// - display denom: eco.<m.CreditTypeAbbrev>.<m.Name>
+func FormatBasketDenom(name, creditTypeAbbrev string, exponent uint32) (string, string, error) {
```
Comment (ryanchristo) [inline x/ecocredit/basket/utils.go:12]: Moved from `msg_create.go` to `utils.go` without changing the underlying logic.

File: x/ecocredit/basket/utils_test.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
+package basket
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestFormatBasketDenom(t *testing.T) {
```
Comment (ryanchristo) [inline x/ecocredit/basket/utils_test.go:9]: Moved from `msg_create_test.go` to `utils_test.go` without changing the test cases.

File: x/ecocredit/server/basket/msg_create.go [hunk 5]
```diff
@@ -16,25 +17,35 @@ import (
 // Create is an RPC to handle basket.MsgCreate
 func (k Keeper) Create(ctx context.Context, msg *basket.MsgCreate) (*basket.MsgCreateResponse, error) {
 	sdkCtx := sdk.UnwrapSDKContext(ctx)
+
 	var fee sdk.Coins
 	k.paramsKeeper.Get(sdkCtx, core.KeyBasketCreationFee, &fee)
-	if err := basket.ValidateMsgCreate(msg, fee); err != nil {
-		return nil, err
+	if !msg.Fee.IsAllGTE(fee) {
+		return nil, sdkerrors.ErrInsufficientFee.Wrapf("minimum fee %s, got %s", fee, msg.Fee)
```
Comment (ryanchristo) [inline x/ecocredit/server/basket/msg_create.go:24]: Moved from `msg_create.go` (stateless message validation) and included directly.

**Comment by ryanchristo** [comment]
@aleem1314 can I get a review when you get a chance? :pray:

**Comment by ryanchristo** [comment]
@aleem1314 or @clevinson when you get a chance :pray:

File: x/ecocredit/server/basket/features/msg_create.feature [hunk 6]
```diff
@@ -1,110 +1,198 @@
...
+  Rule: The user token balance is updated when the basket is created with a fee
+
+    Background:
+      Given a credit type
+      And a minimum basket fee "20regen"
+      And alice has a token balance "40regen"
+
+    Scenario Outline: user token balance is updated
+      When alice attempts to create a basket with fee "<basket-fee>"
+      Then expect the token balance "<token-balance>"
```
Comment (clevinson) [inline x/ecocredit/server/basket/features/msg_create.feature:175]:
```suggestion
      Then expect alice token balance "<token-balance>"
```
possible upgrade?

File: x/ecocredit/server/basket/features/msg_create.feature [hunk 7]
```diff
@@ -1,110 +1,198 @@
...
+      Then expect the error "exponent 3 must be >= credit type precision 6: invalid request"
+
+  Rule: The user token balance is updated when the basket is created with a fee
+
+    Background:
+      Given a credit type
+      And a minimum basket fee "20regen"
+      And alice has a token balance "40regen"
+
+    Scenario Outline: user token balance is updated
```
Comment (clevinson) [inline x/ecocredit/server/basket/features/msg_create.feature:173]:
```suggestion
    Scenario Outline: user token balance is updated and only the required fee is taken
```

File: x/ecocredit/server/basket/features/msg_create.feature [hunk 8]
```diff
@@ -1,110 +1,198 @@
...
+        | description  | basket-fee | token-balance |
+        | greater than | 40regen    | 20regen       |
+        | equal to     | 20regen    | 20regen       |
+
+    # no failing scenario - state transitions only occur upon successful message execution
+
+  Rule: The message response includes basket denom when credits are put into the basket
+
+    Background:
+      Given a credit type
```
Comment (clevinson) [inline x/ecocredit/server/basket/features/msg_create.feature:187]:
```suggestion
      Given a credit type with abbreviation "C" and precision "6"
```

**Review by clevinson — APPROVED** [review]
💯 woooooo!!!

Few nits, but generally great work :)

**Pull request merged**